### PR TITLE
Adding a fallbackLocale to RRule::humanReadable()

### PR DESCRIPTION
**Pull request by gustavs-gutmanis**
It's worth to note that the test is quite worthless if `Intl` is loaded or the `setlocale()` returns actual usable results.

Also `setlocale(LC_ALL, 0)` returns `C/en_US.utf-8/C/C/C/C` for me, and I noticed that it's different per OS and server configurations. So I opted to extract the first matching locale from that string. Do you think that's ok?

### CODE REVIEW

File: src/RRule.php [hunk 1]
```diff
@@ -2324,4 +2339,27 @@ public function humanReadable(array $opt = array())
...
+	 * @return null|string
+	 */
+	private function getSystemLocale()
+	{
+		$locale = null;
+
+		if ( self::$intl_loaded ) {
+			$locale = \Locale::getDefault();
+		} else {
+			$locales = setlocale(LC_ALL, 0);
```
Comment (rlanvin) [inline src/RRule.php]: You should use `setlocale(LC_MESSAGES, 0)`, so you get the locale right away and you don't need to do a `preg_match`.

File: src/RRule.php [hunk 2]
```diff
@@ -2016,11 +2016,21 @@ static protected function i18nList(array $array, $and = 'and')
...
 	 * So region-specific translation files don't need to redefine every strings.
+	 *
+	 * @param string $locale
+	 * @param string $fallbackLocale
+	 *
+	 * @return array
+	 * @throws \InvalidArgumentException
 	 */
-	static protected function i18nLoad($locale)
+	static protected function i18nLoad($locale, $fallbackLocale = 'en_US')
```
Comment (rlanvin) [inline src/RRule.php]: I write the method names in camelCase, but the functions and variables in snake_case, so you should rename this $fallback_locale (or just $fallback is you don't like snake_case)

File: src/RRule.php [hunk 3]
```diff
@@ -2016,11 +2016,21 @@ static protected function i18nList(array $array, $and = 'and')
...
 	{
-		if ( ! preg_match('/^([a-z]{2})(_[A-Z]{2})?$/', $locale, $matches) ) {
-			throw new \InvalidArgumentException('The locale option does not look like a valid locale: '.$locale);
+		if ( ! preg_match('/^([a-z]{2})(_[A-Z]{2})?(_[A-Z]*)?$/', $locale, $matches) ) {
+			if ( $fallbackLocale ) {
+				$locale = $fallbackLocale;
```
Comment (rlanvin) [inline src/RRule.php]: I'm not sure this test should be here. You should use the fallback only when the locale *file* doesn't exist. However, if you pass a wrong locale to the function (such as "foobar"), it should be strict and throw an `invalidArgumentException`.

File: src/RRule.php [hunk 4]
```diff
@@ -2054,15 +2064,20 @@ static protected function i18nLoad($locale)
...
 	 * Format a rule in a human readable string
 	 * intl extension is required.
+	 *
+	 * @param array  $opt
+	 * @param string $fallbackLocale
+	 *
+	 * @return string
 	 */
-	public function humanReadable(array $opt = array())
+	public function humanReadable(array $opt = array(), $fallbackLocale = 'en_US')
```
Comment (rlanvin) [inline src/RRule.php]: You don't need to add a new argument here, it should be part of the `$opt` array.

**Comment by rlanvin** [comment]
Hi,

Thank you for this. There are a couple of things that need to be fixed. I added my comments to your commit. In addition, a couple of notes:

1) What you want is i18nLoad to have the following behavior:
  - If invalid (as in "not formatted correctly") locale, throw an InvalidArgumentException. E.g. if I pass "foobar" as locale
  - If valid (but non existant locale) and invalid fallback, also throw an InvalidArgumentException
  - If valid locale and fallback (but non exitant files for both), throw an InvalidArgumentException.

So what I would do is a little recursivity:
  - test the format of $locale argument
  - load the file
  - if empty result:
    - if there is a fallback, call i18nLoad again, with the fallback locale this time (and no fallback)
    - If no fallback, throw the InvalidArgumentException like now.

2) Your test should reflect that behavior.
  - So you need a test that makes sure invalid locale (such as "foobar") passed to the function will throw an InvalidArgumentException.
  - You need a test that makes sure valid (but strange) locales such as "en_US_POSIX" or "C" do not throw an error.
  - You need a test that makes sure invalid fallback end up throwing an error
  - You need a test that makes sure fallback file is being used when locale file does not exist

It can be only one "testI18nLoad" with many assertions, or multiple different tests (a dataProvider would be useful I guess), that's up to you.

3) I am not sure if the method `getSystemLocale()` is necessary. To me it sounds a bit overkill to have a method just for a "if". More importantly, currently if this method is called before `intl_loaded` is initiated, it will return the wrong locale. What could be more interesting (maybe), it to have there entire `$opt` building process (initializing default values, merging, initializing the date formatter) in a separate method (e.g. `processHumanReadableOptions` or something). Basically, everything that depends on `intl_loaded`, before calling i18nLoad.

Then you could write a test that would do the following:
  - set the locale with setlocale to one thing (e.g. `fr`)
  - set the locale with Locale to another thing (e.g. `en`) (to be skipped in intl is not loaded in the test env)
  - force intl_loaded to false, and assert that locale is `fr`
  - force intl_loaded to true, and assert that locale is `en`

It might be a bit too much though, so I don't mind if you skip that and do everything into `humanReadable` as it is now.

File: src/RRule.php (hunk 1 above)
Comment (gustavs-gutmanis) [inline src/RRule.php]: The problem with `setlocale()` is that it returns `en_US.utf-8` as a locale, so I still need to grep out the `en_US` part. It's possible to make `i18nLoad()`'s preg_match a little bit looser, so that it allows for such locales, since it still extracts what it needs from the locale string.

**Comment by gustavs-gutmanis** [comment]
Thanks for the input, it makes much more sense. Hope I got it right this time.
I set the `$intl_loaded` to `false` during the `humanReadable()` tests, to force the usage of `setlocale()`, and I force the result of `setlocale()` to be `C`, since that is what my system spits out, and then I test some of the edge cases we have gotten to know and love.

There are two tests - one for valid fallback locales and one for invalid ones and a `null` fallback locale, which means that it uses the exact functionality you had before.

Also I loosened up the regex pattern used in `i18nLoad()` to support locales like `en_US.utf-8` which are returned by `setlocale()`.

What do you think about having another test which uses `setlocale()` and `null` fallback locale to test if `setlocale()` works? It's kind of redundant in my opinion, since the fallback locale tests already test the very same functionality.

File: CHANGELOG.md [hunk 5]
```diff
@@ -4,6 +4,7 @@
 
 ### Added
 
+- Support for fallback locale when using `RRule::humanReadable()` [#11](https://github.com/rlanvin/php-rrule/pull/11)
```
Comment (gustavs-gutmanis) [inline CHANGELOG.md]: Please help me out with the `CHANGELOG.md`, is this the correct place to put this string? Should a new minor version title be created? I guess you'd handle that.

File: src/RRule.php [hunk 6]
```diff
@@ -2018,19 +2018,17 @@ static protected function i18nList(array $array, $and = 'and')
...
+	static protected function i18nLoad($locale, $fallback = null)
 	{
-		if ( ! preg_match('/^([a-z]{2})(_[A-Z]{2})?(_[A-Z]*)?$/', $locale, $matches) ) {
-			if ( $fallbackLocale ) {
-				$locale = $fallbackLocale;
-			} else {
-				throw new \InvalidArgumentException('The locale option does not look like a valid locale: '.$locale);
+		if ( ! preg_match('/^([a-z]{2})(_[A-Z]{2})?(?:_[A-Z]*)?/', $locale, $matches) ) {
+			if (!is_null($fallback)) {
+				return self::i18nLoad($fallback);
```
Comment (rlanvin) [inline src/RRule.php]: What is the reason for this? I think when the $locale argument is malformed, it shouldn't try the fallback, it should throw an InvalidArgumentException. For example: `i18nLoad('foobar','en')` should not work in my opinion.

File: src/RRule.php [hunk 7]
```diff
@@ -2066,19 +2067,24 @@ static protected function i18nLoad($locale, $fallbackLocale = 'en_US')
...
 	{
 		if ( self::$intl_loaded === null ) {
 			self::$intl_loaded = extension_loaded('intl');
 		}
 
+		$locale = setlocale(LC_MESSAGES, 0);
+		if ( self::$intl_loaded ) {
+			$locale = \Locale::getDefault();
+		}
+
```
Comment (rlanvin) [inline src/RRule.php]: Here I would maybe handle the case of the special locale "C". So if locale == C, then locale becomes "en" for example.

File: tests/RRuleTest.php [hunk 8]
```diff
@@ -1945,11 +1945,70 @@ public function testDateTimeMutableReferenceBug()
...
+	 * Providing a set of invalid locales to call RRule::i18nLoad() with
+	 *
+	 * @return array
+	 */
+	public function invalidFallbackLocales()
+	{
+		return array(
+			array('eng'),
+			array('invalid'),
+			array('it'),
```
Comment (rlanvin) [inline tests/RRuleTest.php]: This will fail once somebody make a PR to add Italian language :) You should use "xx" instead.

**Comment by rlanvin** [comment]
Great job, it looks much better now! I still have a few minor comments which I added to your code, but overall it's good.

> Also I loosened up the regex pattern used in i18nLoad() to support locales like en_US.utf-8 which are returned by setlocale().

Yes very good catch, I didn't remember setlocale returned locales with a ".utf8".

> Please help me out with the CHANGELOG.md, is this the correct place to put this string?

Yes. You could also add a new line in a section "Fixed" which says "Fixed fatal error Locale class not found when intl extension is not loaded" or something, since your pull request fixes this bug too.

> Should a new minor version title be created? I guess you'd handle that.

I'll handle it once a package a new release. At the moment leave the changes in "Unreleased".

> What do you think about having another test which uses setlocale() and null fallback locale to test if setlocale() works? It's kind of redundant in my opinion, since the fallback locale tests already test the very same functionality.

I'm not sure I understand this question. It seems that you are already doing this test in `testHumanReadable`?

Once you have made the changes, could you please squash all your commits into one, so I can accept the PR? (You will need `git rebase -i` and force push).

Thanks again

File: src/RRule.php (hunk 6 above)
Comment (gustavs-gutmanis) [inline src/RRule.php]: I opted to leave in the 'C' locale and instead of having a special case for it - I thought that it should fail here and try the fallback one. I'll fix this in a sec.

File: src/RRule.php (hunk 7 above)
Comment (gustavs-gutmanis) [inline src/RRule.php]: I made it so at first, but decided to go another route. I'll change it back.

File: tests/RRuleTest.php (hunk 8 above)
Comment (gustavs-gutmanis) [inline tests/RRuleTest.php]: You are totally right :D

File: src/RRule.php (hunk 6 above)
Comment (gustavs-gutmanis) [inline src/RRule.php]: Take a look at [this](http://stackoverflow.com/a/3191729/561378). There are some weird locales out there. `tzm-Latn-DZ` for instance. It would break the locale regex.

**Comment by gustavs-gutmanis** [comment]
I just found out that it's easy to override the locale and not rely on `setlocale()` :D so I updated the tests, and now they test both the `locale` and `fallback`.

I'll update the changelog and squash the commits now.

**Comment by gustavs-gutmanis** [comment]
I hope I did the squashing right.

**Comment by rlanvin** [comment]
> I hope I did the squashing right.

Yes perfect. Thank you!

**Comment by gustavs-gutmanis** [comment]
Sweet!
